### ISSUE TITLE
[모델학습] training dataset 구하기

### ISSUE THREAD

**Issue by Yeonji-Lim** [DGU-Dalcom-2022/superstar-auto-notes-generator#3]
`training/train.py` 를 돌릴 수 있는 dataset 구하기

**Comment by Yeonji-Lim** [comment]
# Deepsaber 측의 설명

`deepsaber-master/README.md`에 의하면 다음과 같다.

하단에 주석 부분을 주석해제하여 내용에 포함함 (숨겨진 내용)

# Getting the data

[TODO] Here we describe the scripts to scrap Beastsaver and BeastSaber to get the training data

### _download data_

`scripts/data_retrieval/download_data.py`

### _obtain the most common states to use for the reduced state representation_

`scripts/data_processing/state_space_functions.py`

# train

## _prepare and preprocess data_

_data augmentation_

`scripts/data_processing/augment_data.sh`

_data preprocessing_

`scripts/feature_extration/process_songs.py` []

## _training_

`scripts/training/script_block_placement.sh`

See more at readme in `scripts/training/README.md`


## Minimal Example of Usage
```python
from base.options.train_options import TrainOptions
from base.data import create_dataset, create_dataloader
from base.models import create_model

opt = TrainOptions().parse()
dataset = create_dataset(opt)
dataloader = create_dataloader(dataset)
data = dataloader[0]
#{'input': songs, 'target': notes, 'features': beat_features}
model = create_model(opt)

for epoch in range(opt.epoch_count, opt.nepoch + opt.nepoch_decay):
    for i, data in enumerate(dataloader)
        model.set_input(data)
        model.optimize_parameters()
        if total_steps % opt.print_freq == 0:
            losses = model.get_current_losses()
            print(losses)
    print(f'End of epoch {i}')
    model.update_learning_rate()
``` 

### Dataset
Steps for creating a custom dataset with custom command line options:
*  Create a file 'datasetname_dataset.py' in the 'data' folder, where datasetname is your custom name.
*  Import the class BaseDataset from base.data.base_dataset
*  Create a class DatasetNameDataset (case does not matter) that inherits from BaseDataset
*  Define the \_\_getitem\_\_ and \_\_len\_\_ methods
*  Options are stored in the opt object that is passed to \_\_init\_\_.
*  Create the 'modify commandline options method' to add your custom commandline options - e.g.:
```python
@staticmethod
def modify_commandline_options(parser, is_train):
    parser.add_argument('--sampling_rate', default=22050, type=float)
    parser.add_argument('--level_diff', default='Expert', help='Difficulty level for beatsaber level')
    parser.add_argument('--hop_length', default=512, type=int)  # Set the hop length; at 22050 Hz, 512 samples ~= 23ms
    parser.add_argument('--compute_feats', action='store_true', help="Whether to extract musical features from the song")
    return parser
```
*  In this example, sampling\_rate, level\_diff, hop\_length and compute\_feats are accessible as attributes of the option object.
*  Add a name() method that returns the name of your class as a string, e.g.
```python
    def name(self):
        return "SongDataset"
```
*  When launching your training script, add the command-line argument --dataset_name=yourDatasetName
*  !!! See base.data.song_dataset.py for an example of a dataset that follows this API !!!

### Model
Steps for creating a custom model with custom command line options:
*  Create a file 'modelname_model.py' in the 'models' folder, where modelname is your custom name.
*  Import the class BaseModel from base.data.base_model
*  Create a class ModelNameModel (case does not matter) that inherits from BaseDataset
*  Initialize a nn.Module instance (your pytorch neural network) with a forward() method, and add it to the ModelNameModel instance as an attribute with name modelname_net
*  Append the module name to the attribute list: self.module\_names.append('modelname') as a string in \_\_init\_\_
*  Add a forward method where the loss is computed. The name of the loss must be self.loss\_lossname.
*  Append lossname to the self.loss\_names list as a string in \_\_init\_\_
*  Add a backward() method where the optimizers are set up, gradients are computed on the losses using self.loss\_lossname.backward(), and an optimizer step is performed.
*  Optimizers cna be stored in the list (self.optimizers)
*  Add an optimize\_parameters() method where self.set\_requires\_grad(self.modelname\_net, requires\_grad=True), self.forward() and self.backward() are defined.
*  Example:
```python
    def __init__(self, opt):
        super().__init__(opt)
        self.loss_names = ['crossentropy']
        self.metric_names = []
        self.module_names = ['wave']
        self.image_paths = []
        self.schedulers = []
        self.net = WaveNet(layers=opt.layers,
                           blocks=opt.blocks,
                           dilation_channels=opt.dilation_channels,
                           residual_channels=opt.residual_channels,
                           skip_channels=opt.skip_channels,
                           end_channels=opt.end_channels,
                           input_channels=opt.input_channels,
                           output_length=opt.output_length,
                           kernel_size=opt.kernel_size,
                           bias=opt.bias)
        self.optimizers = [torch.optim.Adam([
            {'params': [param for name, param in self.net.named_parameters() if name[-4:] == 'bias'],
             'lr': 2 * opt.learning_rate},  # bias parameters change quicker - no weight decay is applied
            {'params': [param for name, param in self.net.named_parameters() if name[-4:] != 'bias'],
             'lr': opt.learning_rate, 'weight_decay': opt.weight_decay}  # filter parameters have weight decay
        ])]

        @staticmethod
    def modify_commandline_options(parser, is_train):
        parser.add_argument('--layers', type=int, default=10, help="Number of layers in each block")
        parser.add_argument('--blocks', type=int, default=4, help="Number of residual blocks in network")
        parser.add_argument('--dilation_channels', type=int, default=32, help="Number of channels in dilated convolutions")
        parser.add_argument('--residual_channels', type=int, default=32, help="Number of channels in the residual link")
        parser.add_argument('--skip_channels', type=int, default=256)
        parser.add_argument('--end_channels', type=int, default=256)
        parser.add_argument('--input_channels', type=int, default=1)
        parser.add_argument('--output_length', type=int, default=1)
        parser.add_argument('--kernel_size', type=int, default=2)
        parser.add_argument('--bias', action='store_false')
        return parser

    def forward(self):
        self.output = self.wave_net.forward(self.input)
        self.loss_crossentropy = F.cross_entropy(self.output, self.target)

    def backward(self):
        self.optimizers[0].zero_grad()
        self.loss_crossentropy.backward()
        self.optimizers[0].step()

    def optimize_parameters(self):
        self.set_requires_grad(self.net, requires_grad=True)
        self.forward()
        self.backward()
        for scheduler in self.schedulers:
            # step for schedulers that update after each iteration
            try:
                scheduler.batch_step()
            except AttributeError:
                pass
```
*  Create the 'modify commandline options method' to add your custom commandline options - (e.g. look above)
*  Add a name() method that returns the name of your class as a string, e.g.
```python
    def name(self):
        return "WaveNetModel"
```
*  !!! See base.models.wavenet_model.py for an example of a dataset that follows this API !!!

# Notes
1.  If the output of your dataset is a dictionary: data = {'input': input\_tensor, 'target': target\_tensor}, you can use model.set_input(data) to store input and target into your model for use in forward.
2.  Store the nn.Module instance in another file (e.g. networks.py) for better abstraction.

**Comment by Yeonji-Lim** [comment]
기존에 사용되는 beat saber의 dataset을 보고, 
슈퍼스타 리소스를 학습에 사용가능한 데이터로 변환하는 방법을 고민하는 것이 좋을 것이라 판단

**Comment by Yeonji-Lim** [comment]
# 기존 데이터 셋 분석

`scripts/data_retrieval/download_data.py`에 의하면, 
[beatsaver.com](https://beatsaver.com/)를 크롤링하여 
임의의 사용자들이 만들 커스텀 데이터들을 수집하여 데이터 셋을 구성하고 있음

## beat saver 각 커스텀 맵 파일 구성
<img width="332" alt="image" src="https://user-images.githubusercontent.com/57888020/182344626-27df75c3-84b9-4d95-b79e-196487fde65f.png">

- 제공하는 난이도 별 .dat 파일
   파일 이름으로 난이도를 구분하고 있음

- 음악(.egg)

- info.dat
   beatsaver.com의 맵 별 정보를 채우기 위한 설정 파일

- 커버이미지(.jpg)

## 난이도 별 .dat 파일 구성

사전에 접해봤던 json 형식의 파일

내용은 다음과 같다.
```json
{
    "_version": "2.0.0",
    "_customData": {
        "_time": 166,
        "_BPMChanges": [
            {
                "_BPM": 250,
                "_time": 12.3125,
                "_beatsPerBar": 4,
                "_metronomeOffset": 4
            }
        ],
        "_bookmarks": [
            {
                "_time": 164.3125,
                "_name": ""
            },
            {
                "_time": 292.3125,
                "_name": ""
            }
        ]
    },
    "_events": [
        {
            "_time": 0,
            "_type": 13,
            "_value": 1
        },
        {
            "_time": 0,
            "_type": 12,
            "_value": 1
        },
        ...
    ],
    "_notes": [
        {
            "_time": 12.3125,
            "_lineIndex": 2,
            "_lineLayer": 0,
            "_type": 1,
            "_cutDirection": 1
        },
        {
            "_time": 15.8125,
            "_lineIndex": 1,
            "_lineLayer": 0,
            "_type": 0,
            "_cutDirection": 1
        },
       ...
   ],
    "_obstacles": [
        {
            "_time": 12.3125,
            "_lineIndex": 0,
            "_type": 0,
            "_duration": 0.25,
            "_width": 1
        },
        {
            "_time": 12.8125,
            "_lineIndex": 0,
            "_type": 1,
            "_duration": 0.125,
            "_width": 1
        },
        ...
    ]
}
```

_events -> 배경이 움직이는 방식

# 기존 코드가 데이터셋을 끌어오지 못한 이유

"http://beatsaver.com/browse/played" 에서 크롤링을 진행하는데, 해당 url 존재 x

**Comment by Yeonji-Lim** [comment]
EXTRACT_DATA 관련 리팩토링할 필요있음 사용되지 않는 데도 계속해서 정의해주고 있음

**Comment by Yeonji-Lim** [comment]
`training/train.py` 에 의하면 `training/data`에서 데이터셋을 가지고 오고 있음

# dataset

그런데 결국 opt.data_dir를 보면 extracted_data에서 가져오고 있다.

```python
# stage_two_dataset.py 53줄 부터,,
if Path(path.parent.__str__()+"/"+diff+".dat").is_file():
       level_file_found = True
```
`.dat` 파일을 가져와서 저장하고 있음

**Comment by Yeonji-Lim** [comment]
이슈 #4 를 참고해서 기존 `~_dataset.py`를 변경하지 않고 구동 가능하도록 json 뿐만아니라 전체 폴더구조도 같게 변경해야 함

**Comment by dolppe** [comment]
BeatSaber의 곡을 다운로드 받아 `data/extraced_data` 에 넣고, `general_beat_saber_dataset.py` 의 init 실행
=>곡 파일인 `.egg`는 전처리 진행 필요
해당 전처리는 `script/feature_extraction/process_songs.py`를 통해 `egg`가 아닌 `npy` 파일로 바뀌어야하는 걸로 보임
넘파이 형식으로 변환 필요

`general_beat_saber_dataset.py` 의 init 실행 중 발생한 에러
```python
features = np.load(features_file)
```
`"An unprocessed song found; need to run preprocessing script process_songs.py before starting to train with them"`

데이터셋을 넣은 형식으로는
`data/extraced_data/`압축 푼 폴더
다운로드 받은 폴더에는 `.egg`  `.dat` 파일로 곡과 노트 파일이 존재 해당 파일 둘 다 모두 이용
다만 `.dat` 파일은 이름 `Expert.dat`와 `ExpertPlus.dat`만 인식하는 것으로 보임(변경 가능, 옵션 파일 이용)

**Comment by Yeonji-Lim** [comment]
# Todo
- [ ] 폴더 구조 일치시키기
- [ ] 음악 전처리
- [ ] .dat 파일, 이름을 난이도로 일치시키기

**Comment by Yeonji-Lim** [comment]
#4 으로 부터 넘겨받은 데이터셋 형식 관련 내용 
(@dolppe 확인 후 추가 내용있다면 정리부탁)

# 데이터셋 전처리 프로세스

1. download_songs로 zip 파일 압축 푼 상태로 extraced_data에 저장
2. process_songs로 extract_feature 적용 해서 np 파일
3. process_songs_tensors로 redueced된 np 파일 (이때 bpm 사용함)
4. 2,3번 파일 가지고 학습

# 요구한 폴더 형식

`data/extraced_data/[a]/[b]`

- a : 곡 이름이나 뭔가 구분지어지는거 다 상관없음 그냥 폴더면 됌
- b : 곡 파일, 전처리된 파일, 난이도.dat 파일 필요

## 파일 명
```
blocks_reduced_classes_file = song_file_path+args.difficulties+"_blocks_reduced_classes_.npy"
features_file = song_file_path+"_"+args.feature_name+"_"+str(args.feature_size)+".npy"
```
